File: leetcode/leetcode_cpp/number-of-provinces.cpp
/*
https://leetcode.com/problems/number-of-provinces

1. DFS
time: o(n)
space: o(n)

2. UnionFind
https://www.youtube.com/watch?v=ayW5B2W9hfo
time: o(logn)
space: o(b)
*/
class Solution {
public:
    bool notVisited(vector<vector<int>>& isConnected, vector<int>& seen, int i) {
        if (seen[i]) return false;
        seen[i] = true;
        for (int j=0; j<isConnected[i].size(); ++j) {
            if (isConnected[i][j])
                notVisited(isConnected, seen, j);
        }
        return true;
    }
    
    int findCircleNum_1(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        int res = 0;
        vector<int> seen(n, 0);
        for (int i=0; i<n; ++i) {
            if (notVisited(isConnected, seen, i))
                res += 1;
        }
        return res;
    }
    
    int find(int i, vector<int>& g) {
        if (g[i] == i) return i;
        return g[i] = find(g[i], g);
    }

    void union_g(int i, int j, vector<int>& g) {
        i = find(i, g);
        j = find(j, g);
        if (i != j)
            g[j] = i;
    }

    int findCircleNum_2(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<int> g(n);
        for (int i=0; i<n; ++i)
            g[i] = i;
        for (int i=0; i<n; ++i) {
            for (int j=i+1; j<n; ++j) {
                if (isConnected[i][j] == 1)
                    union_g(i, j, g);
            }
        }
        int res = 0;
        for (int i=0; i<n; ++i) {
            if (g[i] == i) res += 1;
        }
        return res;
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        return findCircleNum_1(isConnected);
    }
};

File: leetcode/leetcode_py/number-of-provinces.py
'''
https://leetcode.com/problems/number-of-provinces

1. DFS
time: o(n)
space: o(n)

2. UnionFind
https://www.youtube.com/watch?v=ayW5B2W9hfo
time: o(logn)
space: o(b)
'''
class Solution:
    def findCircleNum_1(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        seen = set()
        def dfs(i):
            if i in seen:
                return False
            seen.add(i)
            for j in range(len(isConnected[i])):
                if isConnected[i][j] == 1:
                    dfs(j)
            return True
        res = 0
        for i in range(n):
            if dfs(i):
                res += 1
        return res

    def findCircleNum_2(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        g = []
        for i in range(n):
            g.append(i)

        def find(i):
            if g[i] == i: return i
            g[i] = find(g[i])
            return g[i]
                    
        def join(i, j):
            i = find(i)
            j = find(j)
            if i != j:
                g[j] = i

        for i in range(n):
            for j in range(n):
                if isConnected[i][j]:
                    join(i, j)

        res = 0
        for i in range(n):
            if g[i] == i:
                res += 1
        return res


    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        return self.findCircleNum_2(isConnected)
        
        